### PR TITLE
Add option to override DROP rules for debugging policy.

### PR DESCRIPTION
**Pull request by fasaxc**
Add a "DropActionOverride" config param, which can be set to "DROP", "LOG-and-DROP", ""ACCEPT or "LOG-and-ACCEPT".

The "LOG-" options use an iptables LOG action to log the packet before either ACCEPTing or DROPping the packet.

### CODE REVIEW

File: calico/felix/plugins/fiptgenerator.py [hunk 1]
```diff
@@ -576,13 +578,29 @@ def drop_rules(self, ip_version, chain_name, rule_spec=None, comment=None,
...
-        parts = [p for p in [ipt_action,
-                             chain_name,
-                             rule_spec,
-                             '--jump DROP',
-                             comment_str]
-                 if p is not None]
-        return [' '.join(parts)]
+        rules = []
+
+        if self.ACTION_ON_DROP.startswith("LOG-"):
```
Comment (matthewdupre) [inline calico/felix/plugins/fiptgenerator.py]: I'm amused that this doesn't use the on_deny parameter, passed in to rule_to_iptables_fragments.

File: calico/felix/plugins/fiptgenerator.py [hunk 2]
```diff
@@ -576,13 +578,29 @@ def drop_rules(self, ip_version, chain_name, rule_spec=None, comment=None,
...
+            # log-and-accept, log-and-drop.
+            log_spec = '--jump LOG --log-prefix "CalicoDrop" --log-level 4'
+            log_rule = " ".join(
+                [p for p in [ipt_action, chain_name, rule_spec, log_spec,
+                             comment_str] if p is not None]
+            )
+            rules.append(log_rule)
+
+        if self.ACTION_ON_DROP.endswith("ACCEPT"):
+            action_spec = "--jump ACCEPT"
```
Comment (matthewdupre) [inline calico/felix/plugins/fiptgenerator.py]: There should be a comment in every overridden rule that says that it's overridden.  This is a nasty option to leave on unintentionally: anything we can do to make it more obvious has to be a good thing.

File: calico/felix/config.py [hunk 3]
```diff
@@ -679,6 +690,12 @@ def _validate_cfg(self, final=True):
                 self.parameters["DefaultEndpointToHostAction"]
             )
 
+        if self.ACTION_ON_DROP not in ("DROP", "LOG-and-DROP", "ACCEPT",
+                                       "LOG-and-ACCEPT"):
+            log.warning("Unknown setting for DropActionOverride setting: %s, "
```
Comment (matthewdupre) [inline calico/felix/config.py]: We should also warn if action-on-drop includes ACCEPT.

**Comment by matthewdupre** [comment]
Is there a way we could implement this using more existing code?

This won't actually work as is, but naively the obvious way to implement this would be to use on_deny, and use #1067's logging support; rather than adding separate rule generation.

File: calico/felix/plugins/fiptgenerator.py (hunk 1 above)
Comment (fasaxc) [inline calico/felix/plugins/fiptgenerator.py]: Removed the on_deny/on_allow params under my other PR.  They were just confusing; whether a particular rule should RETURN or ACCEPT, for example, is an implementation detail of the ipt generator.  It doesn't need to be passed in.

**Comment by fasaxc** [comment]
In testing, I found that one of the methods in the iptables generator was bypassing drop_rules().  Now fixed.  I haven't found any other places that generate their own drop rules.

**Pull request merged**